# <center>VAI Academy</center>
# <center> Introdução à Python e Pandas - Parte 7</center>
___
Todo o conteúdo que você terá acesso ao longo desse período é confidencial, não sendo possível compartilhar ou comercializar os links ou os materiais recebidos que sejam de propriedade da VAI Academy. 

Dessa forma, ao participar do curso você está aceitando os termos de confidencialidade e não-comercialização dos conteúdos que serão recebidos.
___

# <center> Objetivos de aprendizado </center>
- Familiarizar-se com as funcionalidades básicas do Pandas
- Ser capaz de carregar dados em um DataFrame
- Ser capaz de realizar manipulações básicas de dados
___

## Conteúdo
1. Introdução (Parte 1)
2. Python/Jupyter Básico (Partes 1 e 2)
3. Python Datatypes (Parte 3)
4. Funções (Parte 4)
5. Numpy (Parte 5)
6. [Pandas (Partes 6 e 7)](#pandas)

In [ ]:
from google.colab import drive
drive.mount('/gdrive')

Mounted at /gdrive


## Carregando Variáveis

Para conseguirmos seguir aprendendo os conceitos a partir de onde paramos no notebook anterior, faremos a leitura de um arquivo pickle na célula a seguir. Não se preocupe em entender o significado da célula a seguir nesse momento, entraremos no detalhe do seu funcionamento mais a frente no curso. O arquivo ```var_AMEM00NB01_parte_7.pkl``` deve se encontrar junto com os arquivos de dados desse notebook. Basta colocá-lo na mesma pasta do seu notebook (caso esteja rodando pelo Jupyter) ou na raiz dos arquivos (caso esteja usando o Colab) e rodar a célula a seguir.

In [ ]:
import pickle

with open("/gdrive/MyDrive/cienciadados/var_AMEM00NB01_parte_7.pkl", "rb") as f:
    pkmn = pickle.load(f)
    fut_players = pickle.load(f)

<a name="pandas"></a>
### 6.4. Juntando DataFrames

É muito comum ter a necessidade de juntar *DataFrames* diferentes. Se você já utilizou SQL ou qualquer outro banco de dados relacional, deve conhecer isso como *join*. O Pandas também tem a mesma função utilizando o método *.merge()*. Antes do exemplo, vamos aprender/relembrar os tipos de *joins* mais comuns:<br>
![Joining Methods](https://i.imgur.com/HaSBT91.jpg) <br>
Agora, vamos carregar um DataFrame mais simples para testar os tipos de *merge*.

In [ ]:
# Execute esta célula para carregar o dataframe metal_bands com dados de bandas de metal
import pandas as pd
metal_bands = pd.read_csv('/gdrive/MyDrive/cienciadados/dados/metal_bands.csv', encoding='latin')
metal_bands.info()
metal_bands.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         5000 non-null   int64 
 1   band_name  5000 non-null   object
 2   fans       5000 non-null   int64 
 3   formed     5000 non-null   object
 4   origin     4992 non-null   object
 5   split      5000 non-null   object
 6   style      5000 non-null   object
dtypes: int64(2), object(5)
memory usage: 273.6+ KB


id    band_name  fans formed          origin split  \
0   0  Iron Maiden  4195   1975  United Kingdom     -   
1   1        Opeth  4147   1990          Sweden  1990   
2   2    Metallica  3712   1981             USA     -   
3   3     Megadeth  3105   1983             USA  1983   
4   4  Amon Amarth  3054   1988          Sweden     -   

                                              style  
0                   New wave of british heavy,Heavy  
1  Extreme progressive,Progressive rock,Progressive  
2                             Heavy,Bay area thrash  
3                            Thrash,Heavy,Hard rock  
4                                     Melodic death

Assim como criamos os dataframes *offensive_stats*, *defensive_stats*, *fire_pkmn* e *water_pkmn*, vamos separar alguns dataframes a partir de *metal_bands* para testar os merges. Observe a célula abaixo.

In [ ]:
bands_origin = metal_bands[['id','band_name','formed','origin']] # ano de formação e país das bandas
bands_style = metal_bands[['id','band_name','style']] # estilo das bandas

bands_split = metal_bands[metal_bands['split']!='-'][['id','band_name','split']] # bandas que se separaram
bands_4000_fans = metal_bands[metal_bands['fans']>4000][['id','band_name','fans']] # bandas com mais de 4000 fans
bands_USA = metal_bands[metal_bands['origin']=='USA'][['id','band_name','formed','origin']] # bandas formadas nos EUA
bands_Sweden = metal_bands[metal_bands['origin']=='Sweden'][['id','band_name','formed','origin']] # bandas formadas na Suécia

Vamos criar um DataFrame a partir de ```bands_origin``` e ```bands_split```, utilizando *merge*.

In [ ]:
origin_split = pd.merge(
    bands_origin, # o DataFrame da esquerda
    bands_split, # o DataFrame da direita
    how='inner', # o tipo de join que queremos fazer
    on='id') # baseado em quais valores em comum (chave)
origin_split.info()
origin_split.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2785 entries, 0 to 2784
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           2785 non-null   int64 
 1   band_name_x  2785 non-null   object
 2   formed       2785 non-null   object
 3   origin       2782 non-null   object
 4   band_name_y  2785 non-null   object
 5   split        2785 non-null   object
dtypes: int64(1), object(5)
memory usage: 152.3+ KB


id    band_name_x formed  origin    band_name_y split
0   1          Opeth   1990  Sweden          Opeth  1990
1   3       Megadeth   1983     USA       Megadeth  1983
2   5         Slayer   1981     USA         Slayer  1981
3   6          Death   1983     USA          Death  2001
4   7  Dream Theater   1985     USA  Dream Theater  1985

Ótimo! Conseguimos fazer o *merge* (termo mais utilizado no Pandas) de dois *DataFrames*. Observe que utilizamos o argumento ```how='inner'```. Lembre-se que *inner*, *left*, *right* e *outer* terão resultados diferentes, observe os merges abaixo e a explicação ao final.

In [ ]:
left_origin_split = pd.merge(bands_origin, bands_split, how='left', on='id')
left_origin_split.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           5000 non-null   int64 
 1   band_name_x  5000 non-null   object
 2   formed       5000 non-null   object
 3   origin       4992 non-null   object
 4   band_name_y  2785 non-null   object
 5   split        2785 non-null   object
dtypes: int64(1), object(5)
memory usage: 273.4+ KB


In [ ]:
right_origin_split = pd.merge(bands_origin, bands_split, how='right', on='id')
right_origin_split.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2785 entries, 0 to 2784
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           2785 non-null   int64 
 1   band_name_x  2785 non-null   object
 2   formed       2785 non-null   object
 3   origin       2782 non-null   object
 4   band_name_y  2785 non-null   object
 5   split        2785 non-null   object
dtypes: int64(1), object(5)
memory usage: 152.3+ KB


In [ ]:
print('Numero de linhas do DataFrame bands_4000_fans:', bands_4000_fans.shape[0])
print('Numero de linhas do DataFrame bands_USA:', bands_USA.shape[0])
print('----------------------------------------------')
outer_origin_split = pd.merge(bands_4000_fans, bands_USA, how='outer', on='id')
outer_origin_split.info()

Numero de linhas do DataFrame bands_4000_fans: 4
Numero de linhas do DataFrame bands_USA: 1139
----------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1143 entries, 0 to 1142
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           1143 non-null   int64  
 1   band_name_x  4 non-null      object 
 2   fans         4 non-null      float64
 3   band_name_y  1139 non-null   object 
 4   formed       1139 non-null   object 
 5   origin       1139 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 62.5+ KB


Como podemos ver com o resultado do método *.info()*, os resultados são de fato bem diferentes.

O *inner* mantém apenas os dados das bandas encontradas nos dois dataframes (onde há correspondência de *id*), dessa forma, a posição do dataframe não faz diferença.

No *left*, mantemos os dados do dataframe à esquerda, e trazemos os dados do dataframe à direita no qual encontrou-se a chave (neste exemplo, o *id* da banda).

Por outro lado, no *right* ocorre o contrário, mantemos os dados do dataframe à direita e, quando há correspondência da chave, trazemos os dados do dataframe à esquerda. Note que o número de entradas (*entries*) é diferente do caso com o *left*. Isso ocorre porque no *left* mantemos os dados de formação das bandas (ou seja, o dataframe contém todas as bandas do .csv), enquanto no *right*, mantemos apenas os dados de bandas que se separaram (e existem muitas bandas que ainda continuam juntas).

Por fim, no *outer* utilizamos dois dataframes diferentes dos anteriores para facilitar o entendimento. Observe pelos prints que existem apenas 4 bandas com mais de 4000 fans e 1139 bandas formadas nos EUA. Quando fazemos o *merge* com *outer*, observe que o total de linhas passa a ser 1143. O que acontece é que esse tipo de join mantém os dados de ambos os dataframes, independente se houve correspondência de chave ou não.

Podemos também querer apenas concatenar dois *DataDrames*, isto é, juntá-los colocando um abaixo ou ao lado do outro. Para isso, utilizamos o método *.concat()*:

In [ ]:
USA_Sweden = pd.concat([bands_USA, bands_Sweden], ignore_index=True) # concatenando bandas formadas nos EUA e bandas formadas na Suécia
USA_Sweden.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1615 entries, 0 to 1614
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         1615 non-null   int64 
 1   band_name  1615 non-null   object
 2   formed     1615 non-null   object
 3   origin     1615 non-null   object
dtypes: int64(1), object(3)
memory usage: 50.6+ KB


Acima fizemos a concatenação vertical. Vamos fazer a horizontal abaixo:

In [ ]:
bands_origin_style = pd.concat([bands_origin, bands_style], axis=1)
bands_origin_style.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         5000 non-null   int64 
 1   band_name  5000 non-null   object
 2   formed     5000 non-null   object
 3   origin     4992 non-null   object
 4   id         5000 non-null   int64 
 5   band_name  5000 non-null   object
 6   style      5000 non-null   object
dtypes: int64(2), object(5)
memory usage: 273.6+ KB


Você deve estar se perguntando: mas então qual a diferença entre utilizar o *merge* e o *concat* com axis=1 (concatenação horizontal)? Observe a imagem abaixo.
![concat](https://i.imgur.com/YlmiwsR.png) <br>

Note que o *concat* recebe os dataframes e apenas os empilha (verticalmente ou horizontalmente). Observe agora o funcionamento do *merge* na imagem abaixo.
![merge](https://i.imgur.com/yGum2id.png) <br>

Com o *merge*, podemos combinar os dataframes de acordo com os valores de suas colunas. Passamos a coluna a ser utilizada como chave, e os valores serão apenas combinados caso haja correspondência nos dois dataframes.

#### Exercício 6.3
Mais uma vez, substitua os \____ de acordo com as instruções

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
fut_players = pd.read_csv('/content/drive/MyDrive/cienciadados/dados/fut_players_data.csv', sep=',')
fut_players.rename(
    columns={'player_id':'id', 'player_name':'name'},
    inplace = True
)
# the_best é um DataDrame dos melhores jogadores em drible (dribbling) e chute (shooting)
the_best = fut_players[(fut_players.dribbling > 90) & (fut_players.shooting > 90)][['id', 'name', 'position', 'dribbling', 'shooting', 'overall']]

# nationalities é um DataDrame da nacionalidade dos jogadores
nationalities = fut_players[['id', 'name', 'nationality']]

# faça um merge dos dois DataDrames, mantendo todos os jogadores de the_best e obtendo suas nacionalidades (dica: a chave é o id)
the_best_nationality = pd.merge(the_best, nationalities, how='left', on='id')
the_best_nationality.head()

id    name_x position  dribbling  shooting  overall    name_y nationality
0   1      Pelé      CAM       96.0      96.0       98      Pelé      Brazil
1   2  Maradona      CAM       97.0      93.0       97  Maradona   Argentina
2   3   Ronaldo       ST       95.0      95.0       96   Ronaldo      Brazil
3   4      Pelé       CF       95.0      93.0       95      Pelé      Brazil
4   5  Maradona      CAM       95.0      91.0       95  Maradona   Argentina

### 6.5. Alterando o dataframe

Até o momento apenas utilizamos os dados da forma que nos foram fornecidos, mas e se precisássemos criar alguma coluna que fosse a combinação das demais? Por exemplo, caso eu deseje criar uma coluna que corresponde à soma do ataque e velocidade dos Pokémons? Observe abaixo:

In [ ]:
# Criando a coluna desejada
pkmn['Sum_Attack_Speed'] = pkmn['Attack'] + pkmn['Speed']
pkmn.head()

#                   Name Type_1  Type_2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp_Atk  Sp_Def  Speed  Generation  Legendary  Sum_Attack_Speed  
0      65      65     45           1      False                94  
1      80      80     60           1      False               122  
2     100     100     80           1      False               162  
3     122     120     80           1      False               180  
4      60      50     65           1      False               117

Observe como foi fácil! Apenas utilizamos o operador de soma com as duas colunas necessárias. Você pode fazer isso com outras operações também, basta utilizar ```-```, ```/``` ou ```*```. Além disso, você pode combinar quantas colunas quiser!

Mas e se precisarmos alterar apenas algumas linhas do nosso DataFrame?

Por exemplo, suponha que você percebeu que seus dados estão errados, e todos os Pokémons com velocidade acima de 100 deveriam estar marcados como Type_1 = 'Fire', podemos seguir o procedimento abaixo:

In [ ]:
# Observe os valores unicos da coluna Type_1 para os Pokémons com mais de 100 de velocidade
pkmn.loc[pkmn['Speed']>100, 'Type_1'].unique()

array(['Bug', 'Normal', 'Electric', 'Ground', 'Psychic', 'Fire', 'Ghost',
       'Water', 'Rock', 'Poison', 'Grass', 'Dark', 'Dragon', 'Steel',
       'Fighting', 'Ice', 'Flying'], dtype=object)

In [ ]:
# Vamos alterar tudo para Fire
pkmn.loc[pkmn['Speed']>100, 'Type_1'] = 'Fire'

In [ ]:
# Observe como os valores mudaram
pkmn.loc[pkmn['Speed']>100, 'Type_1'].unique()

array(['Fire'], dtype=object)

Antes de continuar, vamos ler novamente os dados de Pokémon, sem essa última alteração. Execute a célula abaixo.

In [ ]:
pkmn = pd.read_csv(
    '/content/drive/MyDrive/cienciadados/dados/pokemon_data.csv', # o caminho para o arquivo que se quer ler
    sep=',') # o caracter utilizado para separar os valores

pkmn.rename(
    columns={'Type 1':'Type_1', 'Type 2':'Type_2', 'Sp. Atk':'Sp_Atk','Sp. Def':'Sp_Def'}, # passando o nome antigo e novo como um dicionário
    inplace = True # algumas operações com Pandas criam uma cópia do DataFrame e não alteram o objeto em si, alteramos isso mudando o parâmetro inplace para verdadeiro
)

### 6.6. Operações em grupo

Com Pandas nós podemos aplicar operações em grupos usando o método *.groupby()*. Ele é muito útil por ser uma forma bem simples de extrair informação de dados agregados. Para utilizá-lo, passamos as colunas nas quais queremos agrupar os dados e a operação que queremos fazer. Para exemplificar, vamos ver quantos Pokémons lendários cada geração tem:

In [ ]:
pkmn.groupby('Generation').Legendary.sum() # fazendo uma soma pois a coluna Legendary é boolean

Generation
1     6
2     5
3    18
4    13
5    15
6     8
Name: Legendary, dtype: int64

Podemos obter um relatório da média de diversas colunas para cada tipo de Pokémon:

In [ ]:
pkmn.groupby('Type_1')[['HP','Attack','Defense','Sp_Atk','Sp_Def','Speed']].mean()

HP      Attack     Defense     Sp_Atk     Sp_Def       Speed
Type_1                                                                       
Bug       56.884058   70.971014   70.724638  53.869565  64.797101   61.681159
Dark      66.806452   88.387097   70.225806  74.645161  69.516129   76.161290
Dragon    83.312500  112.125000   86.375000  96.843750  88.843750   83.031250
Electric  59.795455   69.090909   66.295455  90.022727  73.704545   84.500000
Fairy     74.117647   61.529412   65.705882  78.529412  84.705882   48.588235
Fighting  69.851852   96.777778   65.925926  53.111111  64.703704   66.074074
Fire      69.903846   84.769231   67.769231  88.980769  72.211538   74.442308
Flying    70.750000   78.750000   66.250000  94.250000  72.500000  102.500000
Ghost     64.437500   73.781250   81.187500  79.343750  76.468750   64.343750
Grass     67.271429   73.214286   70.800000  77.500000  70.428571   61.928571
Ground    73.781250   95.750000   84.843750  56.468750  62.750000   63.906250
Ice       72.000000   72.750000   71.416667  77.541667  76.291667   63.458333
Normal    77.275510   73.469388   59.846939  55.816327  63.724490   71.551020
Poison    67.250000   74.678571   68.821429  60.428571  64.392857   63.571429
Psychic   70.631579   71.456140   67.684211  98.403509  86.280702   81.491228
Rock      65.363636   92.863636  100.795455  63.340909  75.477273   55.909091
Steel     65.222222   92.703704  126.370370  67.518519  80.629630   55.259259
Water     72.062500   74.151786   72.946429  74.812500  70.517857   65.964286

Isso é realmente muito importante e extremamente utilizado com pandas pois conseguimos fazer análises dos grupos com apenas uma linha de código. Podemos perceber, por exemplo, que Pokémons do tipo *Flying* são especialistas em velocidade enquanto *Dragon* e *Fighting* são especialistas em ataque.

#### Exercício 6.4
Use o método *.groupby()* para descobrir qual país tem o melhor *overall* médio.

In [ ]:
# crie o DataFrame country_avg_overall, que tem o overall médio de cada país (nationality), usando groupby
country_avg_overall = fut_players.groupby('nationality')[['overall']].mean()

# usamos o método idxmax() para encontrar o maior overall médio
print("Melhor overall médio: \n", country_avg_overall.loc[country_avg_overall.idxmax()])
print("Overall médio do Brasil: ", country_avg_overall.loc["Brazil"])

Melhor overall médio: 
                     overall
nationality                
Dominican Republic     79.0
Overall médio do Brasil:  overall    75.219156
Name: Brazil, dtype: float64


### 6.7. Aplicando funções no Pandas

Com Pandas, nós temos um grande nível de controle de nossos dados, e somos capazes de transformá-los conforme precisarmos. Nós podemos, até mesmo, executar funções em DataFrames e manipulá-los como quisermos. Vamos revisitar o método head():

In [ ]:
pkmn.head()

#                   Name Type_1  Type_2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp_Atk  Sp_Def  Speed  Generation  Legendary  
0      65      65     45           1      False  
1      80      80     60           1      False  
2     100     100     80           1      False  
3     122     120     80           1      False  
4      60      50     65           1      False

Existem algumas mega evoluções misturadas no dataset (apenas alguns Pokémons são capazes de evoluir temporariamente para sua forma Mega, uma forma mais poderosa). Não seria legal se nós tivéssemos alguma flag que nos diria se um pokémon é mega ou não? E, por um acaso, será que os pokémons mega são mais poderosos?

Você deve ter percebido que evoluções mega têm um padrão em nosso DataFrame, algo como 'PokemonMega Pokemon'. Se nós tivermos esse padrão, podemos construir uma função que retorna True se este padrão for detectado:

In [ ]:
def is_it_mega(pokemon_name):
    """
    Recebe um nome de pokemon e diz se é uma mega evolução ou não
    I: string pokemon_name
    O: boolean para Mega evos
    """
    if 'Mega ' in pokemon_name: # é importante usar 'Mega ' e não 'mega', pois há um pokemon chamado Yanmega e outro chamado Meganium que não são uma mega evolução
        return True
    else:
        return False

Vamos ver se funciona:

In [ ]:
is_it_mega('VenusaurMega Venusaur')

True

In [ ]:
is_it_mega('Squirtle')

False

Excelente! Seria ótimo se conseguíssemos aplicar essa função em todo nosso DataFrame. Para fazer isso, usaremos o método .apply(). Também criaremos uma coluna que é uma flag se o pokémon é mega:

In [ ]:
pkmn['Mega'] = pkmn.apply(
    lambda row: is_it_mega(row['Name']), # chamando uma função lambda que acabamos de construir
    axis=1 # qual direção queremos executar a função. 0 para horizontal, 1 para vertical
)

In [ ]:
pkmn.head()

#                   Name Type_1  Type_2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp_Atk  Sp_Def  Speed  Generation  Legendary   Mega  
0      65      65     45           1      False  False  
1      80      80     60           1      False  False  
2     100     100     80           1      False  False  
3     122     120     80           1      False   True  
4      60      50     65           1      False  False

Observe como funcionou o apply com lambda. Utilizando o axis=1, a função ```is_it_mega``` é aplicada para cada linha do nosso DataFrame, recebendo como entrada o *Name* do pokémon daquela linha e retornando a flag de True/False na coluna *Mega*.

Agora, vamos verificar quão poderosos são os pokémons mega:

In [ ]:
pkmn.groupby('Mega').Total.mean()

Mega
False    423.457447
True     617.541667
Name: Total, dtype: float64

Uau! Eles têm quase 200 stat points a mais que pokémons normais! Evoluções mega são, sim, muito poderosos! Uma boa prática é sempre tentar manter nosso DataFrame organizado. A forma como os pokémons mega estão nomeados não é muito ótima, e nós já temos uma coluna com a flag para pokémons Mega, então, vamos atacar isso! A estrutura do nome de um pokémon mega é da seguinte forma: 'NomeMega Nome'. Portanto, se nós pegarmos o qeu vem após o caractere ' ', teremos o nome original do pokémon!

In [ ]:
pkmn.Name.nunique() # conta elementos únicos de uma determinada coluna

800

In [ ]:
def get_original_name(s):
    """
    Recebe um nome de pokemon e retorna seu nome original
    I: s string
    O: string
    """
    return s.split(' ')[-1]

pkmn['Name'] = pkmn.Name.apply(lambda s: get_original_name(s)) # sobreescrevendo a coluna Name
pkmn.Name.nunique()

715

In [ ]:
pkmn.head()

#        Name Type_1  Type_2  Total  HP  Attack  Defense  Sp_Atk  Sp_Def  \
0  1   Bulbasaur  Grass  Poison    318  45      49       49      65      65   
1  2     Ivysaur  Grass  Poison    405  60      62       63      80      80   
2  3    Venusaur  Grass  Poison    525  80      82       83     100     100   
3  3    Venusaur  Grass  Poison    625  80     100      123     122     120   
4  4  Charmander   Fire     NaN    309  39      52       43      60      50   

   Speed  Generation  Legendary   Mega  
0     45           1      False  False  
1     60           1      False  False  
2     80           1      False  False  
3     80           1      False   True  
4     65           1      False  False

Agora nós já cobrimos toda a parte básica de Pandas! Vamos praticar essa última parte!

#### Exercício 6.5
Crie uma função que retorna a classificação para o jogador de acordo com as instruções abaixo, então aplique isso para o dataframe fut_players.

In [ ]:
def get_classification(overall):
    """
    Recebe um overall de algum jogador e retorna a classificação conforme a seguir:
    Overall -> classification
    -50     -> "Amador"
    51-60   -> "Ruim"
    61-70   -> "Ok"
    71-80   -> "Bom"
    81-90   -> "Ótimo"
    91+     -> "Lenda"
    
    I: int overall
    O: string
    """
    if (overall >= 91):
      return 'Lenda'
    elif (overall >= 81):
      return 'Ótimo'
    elif (overall >= 71):
      return 'Bom'
    elif (overall >= 61):
      return 'Ok'
    elif (overall >= 51):
      return 'Ruim'
    else:
      return 'Amador'

fut_players['classification'] = fut_players.apply(
    lambda row : get_classification(row['overall']), 
    axis=1
)
fut_players.groupby('classification')['overall'].count()

classification
Amador     135
Bom       4785
Lenda      141
Ok        9500
Ruim      2788
Ótimo     1482
Name: overall, dtype: int64

# Declaração de Inexistência de Plágio:

1. Eu sei que plágio é utilizar o trabalho de outra pessoa e apresentar como meu.
2. Eu sei que plágio é errado e declaro que este notebook foi feito por mim.
3. Tenho consciência de que a utilização do trabalho de terceiros é antiético e está sujeito à medidas administrativas.
4. Declaro também que não compartilhei e não compartilharei meu trabalho com o intuito de que seja copiado e submetido por outra pessoa.

In [ ]:
# LEMBRE-SE DE SALVAR O NOTEBOOK ANTES DE EXECUTAR ESSA CELULA
token = '___' # seu token aqui

# Não altere o código abaixo
import requests as req
exec(req.get('https://api.vai.academy/submissioncode2').text)

# Fim da aula!

Obrigado por participar do curso, você acaba de finalizar a aula de Python, NumPy e Pandas. Neste momento você já deve ser capaz de manipular seus dados no Python, utilizando as bibliotecas que acabamos de aprender! 

Lembre-se que sempre que surgir alguma dúvida, você pode olhar a documentação do [NumPy](https://numpy.org/doc/) e do [Pandas](https://pandas.pydata.org/docs/).